pkg/hive/hive_test.go

```diff
@@ -0,0 +1,217 @@
+// Copyright 2020 The Swarm Authors. All rights reserved.
+// Use of this source code is governed by a BSD-style
+// license that can be found in the LICENSE file.
+
+package hive_test
+
+import (
+	"bytes"
+	"context"
+	"fmt"
+	"io/ioutil"
+	"math/rand"
+	"reflect"
+	"sort"
+	"strconv"
+	"testing"
+	"time"
+
+	ma "github.com/multiformats/go-multiaddr"
+
+	"github.com/ethersphere/bee/pkg/addressbook/inmem"
+	"github.com/ethersphere/bee/pkg/discovery"
+	"github.com/ethersphere/bee/pkg/hive"
+	"github.com/ethersphere/bee/pkg/hive/pb"
+	"github.com/ethersphere/bee/pkg/logging"
+	"github.com/ethersphere/bee/pkg/p2p/protobuf"
+	"github.com/ethersphere/bee/pkg/p2p/streamtest"
+	"github.com/ethersphere/bee/pkg/swarm"
+)
+
+type AddressExporter interface {
+	Overlays() []swarm.Address
+	Multiaddresses() []ma.Multiaddr
+}
+
+func TestBroadcastPeers(t *testing.T) {
+	rand.Seed(time.Now().UnixNano())
+	logger := logging.New(ioutil.Discard, 0)
+
+	// populate all expected and needed random resources for maximum number of messages needed for a test case
+	// tests cases that uses fewer resources can use sub-slices of this data
+	var multiaddrs []ma.Multiaddr
+	var addrs []swarm.Address
+	var records []discovery.BroadcastRecord
+	var wantMsgs []pb.Peers
+
+	// 2*maxBatchSize is ok to be used in the 2 batches corner-case test
+	for i := 0; i < 2*hive.MaxBatchSize; i++ {
+		ma, err := ma.NewMultiaddr("/ip4/127.0.0.1/udp/" + strconv.Itoa(i))
+		if err != nil {
+			t.Fatal(err)
+		}
+
+		multiaddrs = append(multiaddrs, ma)
+		addrs = append(addrs, swarm.NewAddress(createRandomBytes()))
+		if i%hive.MaxBatchSize == 0 {
```

Comment:
My reasoning was that a caller of `hive.BroadcastPeers` could provide any nymber of peers for broadcast. Based on the book, and previous hive implementation I put the maximum batch size for hive to be 50. So basically, if you checkout the implementation of this function, there will be a message per batch. And single batch can contain up to 50 addresses. This is why in this place I need to create a new array (slice) of peers for each batch, so I can append elements to it, to avoid nil pointer exception at ln 60.

Reply:
Done, let me know if it's better